File: lib/batch-factory/director-output.js
const ALLOWED_ASPECT_RATIOS = new Set(['9:16', '16:9']);

function cleanText(value) {
  return typeof value === 'string' ? value.trim() : '';
}

function toInteger(value, field) {
  const number = Number(value);
  if (!Number.isInteger(number)) throw new Error(`${field} 必须是整数`);
  return number;
}

function parseDirectorJson(value) {
  if (value && typeof value === 'object' && !Array.isArray(value)) return value;
  const text = cleanText(value);
  if (!text) throw new Error('导演模型返回为空');
  const fenced = text.match(/```(?:json)?\s*([\s\S]*?)```/i);
  const candidate = fenced ? fenced[1].trim() : text;
  try {
    return JSON.parse(candidate);
  } catch (_) {
    const start = candidate.indexOf('{');
    const end = candidate.lastIndexOf('}');
    if (start !== -1 && end > start) return JSON.parse(candidate.slice(start, end + 1));
    throw new Error('导演模型没有返回合法 JSON');
  }
}

function normalizeNamedPrompts(value, label) {
  const seen = new Set();
  return (Array.isArray(value) ? value : []).map((item, index) => {
    if (!item || typeof item !== 'object' || Array.isArray(item)) throw new Error(`${label}[${index}] 格式无效`);
    const name = cleanText(item.name || item.名称 || item.角色名称 || item.场景名称);
    const prompt = cleanText(item.prompt || item.提示词 || item.description || item.描述);
    if (!name) throw new Error(`${label}[${index}] 缺少名称`);
    if (!prompt) throw new Error(`${label}[${index}] 缺少提示词`);
    if (seen.has(name)) throw new Error(`${label}存在重复名称：${name}`);
    seen.add(name);
    return { name, prompt };
  });
}

function normalizeShots(value, durationSec, videoIndex) {
  const shots = Array.isArray(value) ? value : [];
  if (!shots.length) throw new Error(`storyboard[${videoIndex}] 至少需要一个镜头`);
  let cursor = 0;
  return shots.map((shot, shotIndex) => {
    if (!shot || typeof shot !== 'object' || Array.isArray(shot)) throw new Error(`storyboard[${videoIndex}].shots[${shotIndex}] 格式无效`);
    const startSec = toInteger(shot.start_sec ?? shot.startSec, `storyboard[${videoIndex}].shots[${shotIndex}].start_sec`);
    const endSec = toInteger(shot.end_sec ?? shot.endSec, `storyboard[${videoIndex}].shots[${shotIndex}].end_sec`);
    if (startSec !== cursor) throw new Error(`storyboard[${videoIndex}] 镜头时间轴必须连续，期望从 ${cursor} 秒开始`);
    if (endSec <= startSec) throw new Error(`storyboard[${videoIndex}].shots[${shotIndex}] 结束时间必须大于开始时间`);
    const description = cleanText(shot.description || shot.desc || shot.画面 || shot.prompt);
    if (!description) throw new Error(`storyboard[${videoIndex}].shots[${shotIndex}] 缺少画面描述`);
    cursor = endSec;
    return {
      start_sec: startSec,
      end_sec: endSec,
      shot_type: cleanText(shot.shot_type || shot.shotType || shot.景别),
      camera: cleanText(shot.camera || shot.运镜),
      description
    };
  }).map((shot, index, normalized) => {
    if (index === normalized.length - 1 && shot.end_sec !== durationSec) {
      throw new Error(`storyboard[${videoIndex}] 最后一个镜头必须结束在 ${durationSec} 秒`);
    }
    return shot;
  });
}

function normalizeNameRefs(value) {
  return [...new Set((Array.isArray(value) ? value : String(value || '').split(/[,，]/))
    .map(item => typeof item === 'string' ? item.trim() : cleanText(item?.name))
    .filter(Boolean))];
}

function normalizeDirectorOutput(rawValue, settings = {}) {
  const raw = parseDirectorJson(rawValue);
  const maxVideoDuration = toInteger(settings.maxVideoDuration, 'maxVideoDuration');
  if (maxVideoDuration < 1 || maxVideoDuration > 60) throw new Error('maxVideoDuration 超出允许范围');
  const fixedSingleVideo = settings.fixedSingleVideo === true;
  const exactDuration = fixedSingleVideo ? toInteger(settings.exactDuration, 'exactDuration') : null;
  if (fixedSingleVideo && (exactDuration < 1 || exactDuration > maxVideoDuration)) throw new Error('exactDuration 必须位于模型最大时长范围内');
  const aspectRatio = cleanText(settings.aspectRatio || '9:16');
  if (!ALLOWED_ASPECT_RATIOS.has(aspectRatio)) throw new Error('aspectRatio 仅支持 9:16 或 16:9');

  const characters = normalizeNamedPrompts(raw.characters, 'characters');
  const scenes = normalizeNamedPrompts(raw.scenes, 'scenes');
  const props = normalizeNamedPrompts(raw.props, 'props');
  const characterNames = new Set(characters.map(item => item.name));
  const sceneNames = new Set(scenes.map(item => item.name));
  const propNames = new Set(props.map(item => item.name));
  const allowedPrefixKeys = new Set(Array.isArray(settings.allowedPrefixKeys) ? settings.allowedPrefixKeys : []);
  const sourceVideos = Array.isArray(raw.storyboard) ? raw.storyboard : [];
  if (!sourceVideos.length) throw new Error('storyboard 不能为空');
  if (fixedSingleVideo && sourceVideos.length !== 1) throw new Error('固定单镜头模式只能输出一个视频单元');

  const storyboard = sourceVideos.map((video, videoIndex) => {
    if (!video || typeof video !== 'object' || Array.isArray(video)) throw new Error(`storyboard[${videoIndex}] 格式无效`);
    const durationSec = toInteger(video.duration_sec ?? video.durationSec, `storyboard[${videoIndex}].duration_sec`);
    if (durationSec < 1 || durationSec > maxVideoDuration) throw new Error(`storyboard[${videoIndex}] 时长必须在 1-${maxVideoDuration} 秒之间`);
    if (fixedSingleVideo && durationSec !== exactDuration) throw new Error(`固定单镜头模式必须严格输出 ${exactDuration} 秒`);

    const characterRefs = normalizeNameRefs(video.characters);
    const propRefs = normalizeNameRefs(video.props);
    const scene = cleanText(typeof video.scene === 'string' ? video.scene : video.scene?.name);
    for (const name of characterRefs) if (!characterNames.has(name)) throw new Error(`storyboard[${videoIndex}] 引用了未知人物：${name}`);
    for (const name of propRefs) if (!propNames.has(name)) throw new Error(`storyboard[${videoIndex}] 引用了未知道具：${name}`);
    if (scene && !sceneNames.has(scene)) throw new Error(`storyboard[${videoIndex}] 引用了未知场景：${scene}`);

    const prefixKey = cleanText(video.prefix_key || video.prefixKey);
    if (allowedPrefixKeys.size && prefixKey && !allowedPrefixKeys.has(prefixKey)) throw new Error(`storyboard[${videoIndex}] 使用了未知前缀类型：${prefixKey}`);
    const videoDesc = cleanText(video.video_desc || video.videoDesc);
    const visualPrompt = cleanText(video.visualPrompt || video.visual_prompt || videoDesc);
    if (!videoDesc) throw new Error(`storyboard[${videoIndex}] 缺少 video_desc`);

    return {
      id: video.id ?? videoIndex + 1,
      scene_id: video.scene_id ?? video.sceneId ?? videoIndex + 1,
      duration_sec: durationSec,
      characters: characterRefs,
      props: propRefs,
      scene,
      prefix_key: prefixKey,
      shots: normalizeShots(video.shots, durationSec, videoIndex),
      video_desc: videoDesc,
      visualPrompt
    };
  });

  return {
    characters,
    scenes,
    props,
    storyboard,
    source_coverage: raw.source_coverage && typeof raw.source_coverage === 'object'
      ? {
        source_complete: raw.source_coverage.source_complete === true,
        source_end_marker: cleanText(raw.source_coverage.source_end_marker),
        has_remaining_source: raw.source_coverage.has_remaining_source === true
      }
      : { source_complete: !fixedSingleVideo, source_end_marker: '', has_remaining_source: fixedSingleVideo }
  };
}

module.exports = {
  ALLOWED_ASPECT_RATIOS,
  parseDirectorJson,
  normalizeDirectorOutput
};

File: backend/internal/batchfactoryv11/director_output.go
package batchfactoryv11

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
)

type DirectorSettings struct {
	MaxVideoDuration    int      `json:"maxVideoDuration"`
	AudioTargetSeconds  int      `json:"audioTargetSeconds,omitempty"`
	FixedSingleVideo    bool     `json:"fixedSingleVideo"`
	AspectRatio         string   `json:"aspectRatio"`
	AllowedPrefixKeys   []string `json:"allowedPrefixKeys"`
	RequireVisualPrompt bool     `json:"requireVisualPrompt"`
	RequireH3Metadata   bool     `json:"requireH3Metadata"`
}

type NamedPrompt struct {
	Name   string `json:"name"`
	Prompt string `json:"prompt"`
}

type DirectorShot struct {
	StartSec      int    `json:"start_sec"`
	EndSec        int    `json:"end_sec"`
	ShotType      string `json:"shot_type"`
	Camera        string `json:"camera"`
	Rhythm        string `json:"rhythm,omitempty"`
	Audio         string `json:"audio,omitempty"`
	VisualContext string `json:"visual_context,omitempty"`
	Lighting      string `json:"lighting,omitempty"`
	Description   string `json:"description"`
}

type DirectorVideo struct {
	ID           any            `json:"id"`
	SceneID      any            `json:"scene_id"`
	DurationSec  int            `json:"duration_sec"`
	Characters   []string       `json:"characters"`
	Props        []string       `json:"props"`
	Scene        string         `json:"scene"`
	SceneMemory  string         `json:"scene_memory,omitempty"`
	PrefixKey    string         `json:"prefix_key"`
	Shots        []DirectorShot `json:"shots"`
	VideoDesc    string         `json:"video_desc"`
	VisualPrompt string         `json:"visual_prompt,omitempty"`
}

type SourceCoverage struct {
	SourceComplete     bool   `json:"source_complete"`
	SourceEndMarker    string `json:"source_end_marker"`
	HasRemainingSource bool   `json:"has_remaining_source"`
}

type DirectorResult struct {
	Characters           []NamedPrompt         `json:"characters"`
	Scenes               []NamedPrompt         `json:"scenes"`
	Props                []NamedPrompt         `json:"props"`
	Storyboard           []DirectorVideo       `json:"storyboard"`
	SourceCoverage       SourceCoverage        `json:"source_coverage"`
	SmartUnifiedStyle    string                `json:"smart_unified_style,omitempty"`
	SmartUnifiedAnalysis *SmartUnifiedAnalysis `json:"smart_unified_analysis,omitempty"`
	H3Director           *H3DirectorDocument   `json:"h3_director,omitempty"`
}

// storyboardVideoPrompt is the durable card body for one VIDEO.  The provider
// returns video_desc as a compact internal summary, but public screenplay
// generation exposes the verified shot timeline as the actual video prompt.
// Keep the summary for search/audit only; never let it replace a card's
// camera, timing and visible-action instructions.
func storyboardVideoPrompt(video DirectorVideo) string {
	if len(video.Shots) == 0 {
		return strings.TrimSpace(video.VideoDesc)
	}
	lines := []string{"镜头画面："}
	for _, shot := range video.Shots {
		lines = append(lines, fmt.Sprintf(
			"%02d:%02d-%02d:%02d | %s｜%s | %s",
			shot.StartSec/60, shot.StartSec%60,
			shot.EndSec/60, shot.EndSec%60,
			strings.TrimSpace(shot.ShotType),
			strings.TrimSpace(shot.Camera),
			strings.TrimSpace(shot.Description),
		))
	}
	return strings.Join(lines, "\n")
}

var fencedDirectorJSON = regexp.MustCompile("(?is)```(?:json)?\\s*([\\s\\S]*?)```")

func ParseDirectorJSON(value string) (json.RawMessage, error) {
	text := strings.TrimSpace(value)
	if text == "" {
		return nil, fmt.Errorf("导演模型返回为空")
	}
	candidate := text
	if match := fencedDirectorJSON.FindStringSubmatch(text); len(match) > 1 {
		candidate = strings.TrimSpace(match[1])
	}
	if json.Valid([]byte(candidate)) {
		return json.RawMessage(candidate), nil
	}
	start := strings.Index(candidate, "{")
	end := strings.LastIndex(candidate, "}")
	if start >= 0 && end > start {
		fragment := candidate[start : end+1]
		if json.Valid([]byte(fragment)) {
			return json.RawMessage(fragment), nil
		}
	}
	return nil, fmt.Errorf("导演模型没有返回合法 JSON")
}

func cleanDirectorText(value any) string {
	if text, ok := value.(string); ok {
		return strings.TrimSpace(text)
	}
	return ""
}

func decodeDirectorObject(raw json.RawMessage) (map[string]any, error) {
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber()
	var value map[string]any
	if err := decoder.Decode(&value); err != nil {
		return nil, fmt.Errorf("导演模型没有返回合法 JSON")
	}
	if value == nil {
		return nil, fmt.Errorf("导演模型没有返回合法 JSON")
	}
	return value, nil
}

func directorInteger(value any, field string) (int, error) {
	var number float64
	switch typed := value.(type) {
	case json.Number:
		parsed, err := strconv.ParseFloat(string(typed), 64)
		if err != nil {
			return 0, fmt.Errorf("%s 必须是整数", field)
		}
		number = parsed
	case float64:
		number = typed
	case float32:
		number = float64(typed)
	case int:
		return typed, nil
	case int64:
		return int(typed), nil
	case string:
		parsed, err := strconv.ParseFloat(strings.TrimSpace(typed), 64)
		if err != nil {
			return 0, fmt.Errorf("%s 必须是整数", field)
		}
		number = parsed
	default:
		return 0, fmt.Errorf("%s 必须是整数", field)
	}
	if math.IsNaN(number) || math.IsInf(number, 0) || math.Trunc(number) != number {
		return 0, fmt.Errorf("%s 必须是整数", field)
	}
	return int(number), nil
}

func directorArray(value any) []any {
	if array, ok := value.([]any); ok {
		return array
	}
	return []any{}
}

func directorMap(value any) (map[string]any, bool) {
	entry, ok := value.(map[string]any)
	return entry, ok
}

func firstDirectorText(entry map[string]any, fields ...string) string {
	for _, field := range fields {
		if text := cleanDirectorText(entry[field]); text != "" {
			return text
		}
	}
	return ""
}

func normalizeDirectorNamedPrompts(value any, label string) ([]NamedPrompt, error) {
	items := directorArray(value)
	seen := map[string]struct{}{}
	out := make([]NamedPrompt, 0, len(items))
	for index, item := range items {
		entry, ok := directorMap(item)
		if !ok {
			return nil, fmt.Errorf("%s[%d] 格式无效", label, index)
		}
		name := firstDirectorText(entry, "name", "名称", "角色名称", "场景名称")
		prompt := firstDirectorText(entry, "prompt", "提示词", "description", "描述")
		if name == "" {
			return nil, fmt.Errorf("%s[%d] 缺少名称", label, index)
		}
		if prompt == "" {
			return nil, fmt.Errorf("%s[%d] 缺少提示词", label, index)
		}
		if _, exists := seen[name]; exists {
			return nil, fmt.Errorf("%s存在重复名称：%s", label, name)
		}
		seen[name] = struct{}{}
		out = append(out, NamedPrompt{Name: name, Prompt: prompt})
	}
	return out, nil
}

func normalizeDirectorRefs(value any) []string {
	values := []any{}
	switch typed := value.(type) {
	case []any:
		values = typed
	case string:
		normalized := strings.ReplaceAll(typed, "，", ",")
		for _, part := range strings.Split(normalized, ",") {
			values = append(values, part)
		}
	default:
		if value != nil {
			values = append(values, value)
		}
	}
	seen := map[string]struct{}{}
	out := make([]string, 0, len(values))
	for _, item := range values {
		name := cleanDirectorText(item)
		if entry, ok := directorMap(item); ok {
			name = cleanDirectorText(entry["name"])
		}
		if name == "" {
			continue
		}
		if _, exists := seen[name]; exists {
			continue
		}
		seen[name] = struct{}{}
		out = append(out, name)
	}
	return out
}

func directorNameSet(values []NamedPrompt) map[string]struct{} {
	out := make(map[string]struct{}, len(values))
	for _, value := range values {
		out[value.Name] = struct{}{}
	}
	return out
}

func normalizeDirectorShots(value any, durationSec, videoIndex int, requireH3Metadata bool) ([]DirectorShot, error) {
	items := directorArray(value)
	if len(items) == 0 {
		return nil, fmt.Errorf("storyboard[%d] 至少需要一个镜头", videoIndex)
	}
	cursor := 0
	out := make([]DirectorShot, 0, len(items))
	for shotIndex, item := range items {
		entry, ok := directorMap(item)
		if !ok {
			return nil, fmt.Errorf("storyboard[%d].shots[%d] 格式无效", videoIndex, shotIndex)
		}
		startValue := entry["start_sec"]
		if startValue == nil {
			startValue = entry["startSec"]
		}
		endValue := entry["end_sec"]
		if endValue == nil {
			endValue = entry["endSec"]
		}
		startSec, err := directorInteger(startValue, fmt.Sprintf("storyboard[%d].shots[%d].start_sec", videoIndex, shotIndex))
		if err != nil {
			return nil, err
		}
		endSec, err := directorInteger(endValue, fmt.Sprintf("storyboard[%d].shots[%d].end_sec", videoIndex, shotIndex))
		if err != nil {
			return nil, err
		}
		if startSec != cursor {
			return nil, fmt.Errorf("storyboard[%d] 镜头时间轴必须连续，期望从 %d 秒开始", videoIndex, cursor)
		}
		if endSec <= startSec {
			return nil, fmt.Errorf("storyboard[%d].shots[%d] 结束时间必须大于开始时间", videoIndex, shotIndex)
		}
		// Providers commonly return an otherwise valid shot with a Chinese field
		// name (or `shot_description`) for its visible action. Keep the persisted
		// contract canonical as `description`, while accepting these equivalent
		// representations at the boundary. We still reject an actually empty shot
		// because a VIDEO cannot be planned from it.
		description := firstDirectorText(entry, "description", "shot_description", "shotDescription", "visual_description", "visualDescription", "desc", "画面描述", "画面内容", "画面", "prompt")
		if description == "" {
			return nil, fmt.Errorf("storyboard[%d].shots[%d] 缺少画面描述", videoIndex, shotIndex)
		}
		rhythm := firstDirectorText(entry, "rhythm", "pace", "节奏")
		audio := firstDirectorText(entry, "audio", "audio_raw", "audioRaw", "sound", "dialogue", "声音", "音频", "对白")
		visualContext := firstDirectorText(entry, "visual_context", "visualContext", "scene_context", "sceneContext", "场景上下文", "场景")
		lighting := firstDirectorText(entry, "lighting", "light", "light_description", "lightDescription", "光影", "光线")
		if requireH3Metadata && rhythm == "" {
			return nil, fmt.Errorf("H3 storyboard[%d].shots[%d] 缺少 rhythm", videoIndex, shotIndex)
		}
		if requireH3Metadata && audio == "" {
			return nil, fmt.Errorf("H3 storyboard[%d].shots[%d] 缺少 audio（无声音时填写“无”）", videoIndex, shotIndex)
		}
		if requireH3Metadata && visualContext == "" {
			return nil, fmt.Errorf("H3 storyboard[%d].shots[%d] 缺少 visual_context", videoIndex, shotIndex)
		}
		if requireH3Metadata && lighting == "" {
			return nil, fmt.Errorf("H3 storyboard[%d].shots[%d] 缺少 lighting", videoIndex, shotIndex)
		}
		cursor = endSec
		out = append(out, DirectorShot{
			StartSec:      startSec,
			EndSec:        endSec,
			ShotType:      firstDirectorText(entry, "shot_type", "shotType", "景别"),
			Camera:        firstDirectorText(entry, "camera", "运镜"),
			Rhythm:        rhythm,
			Audio:         audio,
			VisualContext: visualContext,
			Lighting:      lighting,
			Description:   description,
		})
	}
	if out[len(out)-1].EndSec != durationSec {
		return nil, fmt.Errorf("storyboard[%d] 最后一个镜头必须结束在 %d 秒", videoIndex, durationSec)
	}
	return out, nil
}

func NormalizeDirectorOutput(raw json.RawMessage, settings DirectorSettings) (DirectorResult, error) {
	if settings.MaxVideoDuration < 1 || settings.MaxVideoDuration > 60 {
		return DirectorResult{}, fmt.Errorf("maxVideoDuration 超出允许范围")
	}
	if settings.AudioTargetSeconds < 0 {
		return DirectorResult{}, fmt.Errorf("audioTargetSeconds 不能小于 0")
	}
	if settings.FixedSingleVideo && settings.AudioTargetSeconds > 0 {
		return DirectorResult{}, fmt.Errorf("固定开头只生产 VIDEO01，不能同时执行配音分镜规划")
	}
	aspectRatio := strings.TrimSpace(settings.AspectRatio)
	if aspectRatio == "" {
		aspectRatio = "9:16"
	}
	if aspectRatio != "9:16" && aspectRatio != "16:9" {
		return DirectorResult{}, fmt.Errorf("aspectRatio 仅支持 9:16 或 16:9")
	}

	root, err := decodeDirectorObject(raw)
	if err != nil {
		return DirectorResult{}, err
	}
	characters, err := normalizeDirectorNamedPrompts(root["characters"], "characters")
	if err != nil {
		return DirectorResult{}, err
	}
	scenes, err := normalizeDirectorNamedPrompts(root["scenes"], "scenes")
	if err != nil {
		return DirectorResult{}, err
	}
	props, err := normalizeDirectorNamedPrompts(root["props"], "props")
	if err != nil {
		return DirectorResult{}, err
	}
	characterNames := directorNameSet(characters)
	sceneNames := directorNameSet(scenes)
	propNames := directorNameSet(props)
	allowedPrefixes := map[string]struct{}{}
	for _, key := range settings.AllowedPrefixKeys {
		allowedPrefixes[key] = struct{}{}
	}

	sourceVideos := directorArray(root["storyboard"])
	if len(sourceVideos) == 0 {
		return DirectorResult{}, fmt.Errorf("storyboard 不能为空")
	}
	if settings.AudioTargetSeconds > 0 {
		minimumVideos := audioMinimumVideoCount(settings.AudioTargetSeconds, settings.MaxVideoDuration)
		if len(sourceVideos) < minimumVideos {
			return DirectorResult{}, fmt.Errorf("配音规划需要至少 %d 个 VIDEO 才能覆盖 %d 秒（单段上限 %d 秒），当前只有 %d 个", minimumVideos, settings.AudioTargetSeconds, settings.MaxVideoDuration, len(sourceVideos))
		}
	}
	storyboard := make([]DirectorVideo, 0, len(sourceVideos))
	for videoIndex, item := range sourceVideos {
		video, ok := directorMap(item)
		if !ok {
			return DirectorResult{}, fmt.Errorf("storyboard[%d] 格式无效", videoIndex)
		}
		durationValue := video["duration_sec"]
		if durationValue == nil {
			durationValue = video["durationSec"]
		}
		durationSec, err := directorInteger(durationValue, fmt.Sprintf("storyboard[%d].duration_sec", videoIndex))
		if err != nil {
			return DirectorResult{}, err
		}
		if durationSec < 1 || durationSec > settings.MaxVideoDuration {
			return DirectorResult{}, fmt.Errorf("storyboard[%d] 时长必须在 1-%d 秒之间", videoIndex, settings.MaxVideoDuration)
		}
		characterRefs := normalizeDirectorRefs(video["characters"])
		propRefs := normalizeDirectorRefs(video["props"])
		scene := cleanDirectorText(video["scene"])
		if sceneEntry, ok := directorMap(video["scene"]); ok {
			scene = cleanDirectorText(sceneEntry["name"])
		}
		for _, name := range characterRefs {
			if _, exists := characterNames[name]; !exists {
				return DirectorResult{}, fmt.Errorf("storyboard[%d] 引用了未知人物：%s", videoIndex, name)
			}
		}
		for _, name := range propRefs {
			if _, exists := propNames[name]; !exists {
				return DirectorResult{}, fmt.Errorf("storyboard[%d] 引用了未知道具：%s", videoIndex, name)
			}
		}
		if scene != "" {
			if _, exists := sceneNames[scene]; !exists {
				return DirectorResult{}, fmt.Errorf("storyboard[%d] 引用了未知场景：%s", videoIndex, scene)
			}
		}

		prefixKey := firstDirectorText(video, "prefix_key", "prefixKey")
		if len(allowedPrefixes) > 0 && prefixKey != "" {
			if _, exists := allowedPrefixes[prefixKey]; !exists {
				return DirectorResult{}, fmt.Errorf("storyboard[%d] 使用了未知前缀类型：%s", videoIndex, prefixKey)
			}
		}
		videoDesc := firstDirectorText(video, "video_desc", "videoDesc")
		if videoDesc == "" {
			return DirectorResult{}, fmt.Errorf("storyboard[%d] 缺少 video_desc", videoIndex)
		}
		visualPrompt := firstDirectorText(video, "visual_prompt", "visualPrompt", "image_prompt", "imagePrompt")
		if settings.RequireVisualPrompt && visualPrompt == "" {
			return DirectorResult{}, fmt.Errorf("storyboard[%d] 缺少 visual_prompt", videoIndex)
		}
		if !settings.RequireVisualPrompt {
			visualPrompt = ""
		}
		shots, err := normalizeDirectorShots(video["shots"], durationSec, videoIndex, settings.RequireH3Metadata)
		if err != nil {
			return DirectorResult{}, err
		}
		sceneMemory := firstDirectorText(video, "scene_memory", "sceneMemory", "continuity_state", "continuityState")
		if settings.RequireH3Metadata && sceneMemory == "" {
			return DirectorResult{}, fmt.Errorf("H3 storyboard[%d] 缺少 scene_memory", videoIndex)
		}

		id := video["id"]
		if id == nil {
			id = videoIndex + 1
		}
		sceneID := video["scene_id"]
		if sceneID == nil {
			sceneID = video["sceneId"]
		}
		if sceneID == nil {
			sceneID = videoIndex + 1
		}
		storyboard = append(storyboard, DirectorVideo{
			ID:           id,
			SceneID:      sceneID,
			DurationSec:  durationSec,
			Characters:   characterRefs,
			Props:        propRefs,
			Scene:        scene,
			SceneMemory:  sceneMemory,
			VisualPrompt: visualPrompt,
			PrefixKey:    prefixKey,
			Shots:        shots,
			VideoDesc:    videoDesc,
		})
	}
	if settings.AudioTargetSeconds > 0 {
		total := 0
		for _, video := range storyboard {
			total += video.DurationSec
		}
		if total != settings.AudioTargetSeconds {
			return DirectorResult{}, fmt.Errorf("所有分镜总时长必须严格等于配音规划时长 %d 秒，当前为 %d 秒", settings.AudioTargetSeconds, total)
		}
	}

	coverage := SourceCoverage{SourceComplete: true, SourceEndMarker: "", HasRemainingSource: false}
	if rawCoverage, ok := directorMap(root["source_coverage"]); ok {
		coverage = SourceCoverage{
			SourceComplete:     rawCoverage["source_complete"] == true,
			SourceEndMarker:    cleanDirectorText(rawCoverage["source_end_marker"]),
			HasRemainingSource: rawCoverage["has_remaining_source"] == true,
		}
	}
	return DirectorResult{
		Characters:     characters,
		Scenes:         scenes,
		Props:          props,
		Storyboard:     storyboard,
		SourceCoverage: coverage,
	}, nil
}

// NormalizeAssetExtractionOutput accepts the compact asset-only result used by
// the per-book asset modal. It deliberately does not inspect storyboard data.
func NormalizeAssetExtractionOutput(raw json.RawMessage) (DirectorAssets, error) {
	root, err := decodeDirectorObject(raw)
	if err != nil {
		return DirectorAssets{}, err
	}
	characters, err := normalizeDirectorNamedPrompts(root["characters"], "characters")
	if err != nil {
		return DirectorAssets{}, err
	}
	scenes, err := normalizeDirectorNamedPrompts(root["scenes"], "scenes")
	if err != nil {
		return DirectorAssets{}, err
	}
	props, err := normalizeDirectorNamedPrompts(root["props"], "props")
	if err != nil {
		return DirectorAssets{}, err
	}
	if len(characters)+len(scenes)+len(props) == 0 {
		return DirectorAssets{}, fmt.Errorf("人物、场景、道具至少应提取一项")
	}
	return DirectorAssets{Characters: characters, Scenes: scenes, Props: props}, nil
}
